ubuntu 22.04 failed to setup depends because package name is wrong

in ubuntu 22.04, 
after I run setup depends scripts, I got below error

there are 4 packages name is not correct under ubuntu 22.04

E: Unable to locate package libbpf-devel
E: Package 'bpftool' has no installation candidate
E: Unable to locate package libblkid-devel
E: Unable to locate package zlib

```
Package bpftool is a virtual package provided by:
  linux-nvidia-6.2-tools-common 6.2.0-1003.3~22.04.1
  linux-lowlatency-hwe-6.5-tools-common 6.5.0-27.28.1~22.04.1
  linux-hwe-6.5-tools-common 6.5.0-27.28~22.04.1
  linux-hwe-6.2-tools-common 6.2.0-39.40~22.04.1
You should explicitly select one to install.

E: Unable to locate package libbpf-devel
E: Package 'bpftool' has no installation candidate
E: Unable to locate package libblkid-devel
E: Unable to locate package zlib
[IOTRACE][ERROR] Cannot install required dependencies
```

--- a/README.md
+++ b/README.md
@@ -47,6 +47,7 @@ following OSes:
 | OS     | Version | Kernel Version |
 | ------ | ------- | -------------- |
 | Fedora | 36      | 5.18.10        |
+| Ubuntu | 22.04   | 5.15.0         |
 
 <a id="source"></a>
 

--- a/setup_dependencies.sh
+++ b/setup_dependencies.sh
@@ -1,6 +1,7 @@
 #!/bin/bash
 
 # Copyright(c) 2012-2020 Intel Corporation
+# Copyright 2024 Solidigm All Rights Reserved
 # SPDX-License-Identifier: BSD-3-Clause
 
 IOTRACE_DIR="$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)"
@@ -25,17 +26,14 @@ function iotrace_get_kernel_package () {
 }
 
 function iotrace_get_distribution_pkg_dependencies () {
-    local pkgs="zlib clang llvm libblkid-devel bpftool libbpf-devel"
+    local pkgs="clang llvm"
 
     case "${DISTRO}" in
-    "RHEL7"|"CENTOS7"|"FEDORA")
-        echo "${pkgs} rpm-build elfutils-libelf-devel"
-        ;;
-    "RHEL8"|"CENTOS8")
-        echo "${pkgs} rpm-build elfutils-libelf-devel"
+    "RHEL7"|"CENTOS7"|"RHEL8"|"CENTOS8"|"FEDORA")
+        echo "${pkgs} rpm-build elfutils-libelf-devel libblkid-devel libbpf-devel zlib bpftool"
         ;;
     "UBUNTU"|"DEBIAN")
-        echo "${pkgs} dpkg"
+        echo "${pkgs} dpkg libblkid-dev libbpf-dev zlib1g-dev linux-tools-common linux-tools-generic"
         ;;
     *)
         error "Unknown Linux distribution"

--- a/source/iotrace/configure.d/1_page.conf
+++ b/source/iotrace/configure.d/1_page.conf
@@ -1,0 +1,34 @@
+#!/bin/bash
+#
+# Copyright 2024 Solidigm All Rights Reserved
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+. $(dirname $3)/conf_framework
+
+check() {
+    cur_name=$(basename $2)
+    config_file_path=$1
+    if compile_check $cur_name "struct folio f = {}"
+    then
+        echo $cur_name "2" >> $config_file_path
+    elif compile_check $cur_name "struct page p = {}"
+    then
+        echo $cur_name "1" >> $config_file_path
+    else
+        echo $cur_name "X" >> $config_file_path
+    fi
+}
+
+apply() {
+    case "$1" in
+    "1")
+        add_define "IOTRACE_PAGE_VERSION 1";;
+    "2")
+        add_define "IOTRACE_PAGE_VERSION 2";;
+    *)
+        exit 1
+    esac
+}
+
+conf_run $@

--- a/source/iotrace/iotrace.bpf.defs.h
+++ b/source/iotrace/iotrace.bpf.defs.h
@@ -1,5 +1,6 @@
 /*
- * Copyright(c) 2012-2020 Intel Corporation
+ * Copyright(c) 2012-2020 Intel
+ * Copyright 2024 Solidigm All Rights Reserved
  * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
@@ -170,6 +171,36 @@ static __always_inline uint64_t iotrace_rq_to_id(const struct request *rq) {
  * Page definitions
  */
 
+static __always_inline int iotrace_page_tail(struct page *page) {
+    return BPF_CORE_READ(page, compound_head) & 1;
+}
+
+static __always_inline bool iotrace_page_slab(struct page *page) {
+    if (BPF_CORE_READ(page, flags) & (1UL < PG_slab)) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+static __always_inline uint64_t iotrace_page_index(struct page *page) {
+    return BPF_CORE_READ(page, index);
+}
+
+static __always_inline int iotrace_page_compound(struct page *page) {
+    if (BPF_CORE_READ(page, flags) & (1UL < PG_head)) {
+        return true;
+    }
+
+    if (iotrace_page_tail(page)) {
+        return true;
+    }
+
+    return false;
+}
+
+#if IOTRACE_PAGE_VERSION > 1
+
 static __always_inline unsigned long _compound_head(const struct page *page) {
     unsigned long head = BPF_CORE_READ(page, compound_head);
 
@@ -197,30 +228,6 @@ static __always_inline int iotrace_page_read_ahead(struct page *page) {
     return BPF_CORE_READ(page, flags) & (1UL < PG_readahead);
 }
 
-static __always_inline int iotrace_page_tail(struct page *page) {
-    return BPF_CORE_READ(page, compound_head) & 1;
-}
-
-static __always_inline int iotrace_page_compound(struct page *page) {
-    if (BPF_CORE_READ(page, flags) & (1UL < PG_head)) {
-        return true;
-    }
-
-    if (iotrace_page_tail(page)) {
-        return true;
-    }
-
-    return false;
-}
-
-static __always_inline bool iotrace_page_slab(struct page *page) {
-    if (BPF_CORE_READ(page, flags) & (1UL < PG_slab)) {
-        return true;
-    } else {
-        return false;
-    }
-}
-
 static __always_inline bool iotrace_page_has_mapping(struct page *page) {
     struct folio *folio = page_folio(page);
 
@@ -234,9 +241,30 @@ static __always_inline struct inode *iotrace_page_inode(struct page *page) {
     return inode;
 }
 
-static __always_inline uint64_t iotrace_page_index(struct page *page) {
-    return BPF_CORE_READ(page, index);
+#else
+
+static __always_inline bool iotrace_page_anon(struct page *page) {
+#define PAGE_MAPPING_ANON 0x1
+    struct address_space *mapping = BPF_CORE_READ(page, mapping);
+
+    return ((unsigned long) mapping & PAGE_MAPPING_ANON) != 0;
 }
+
+static __always_inline int iotrace_page_read_ahead(struct page *page) {
+    return 0;
+}
+
+static __always_inline bool iotrace_page_has_mapping(struct page *page) {
+    return NULL != BPF_CORE_READ(page, mapping);
+}
+
+static __always_inline struct inode *iotrace_page_inode(struct page *page) {
+    struct inode *inode = BPF_CORE_READ(page, mapping, host);
+
+    return inode;
+}
+
+#endif
 
 /*
  * Inode definitions

